src/librustc_mir/transform/qualify_min_const_fn.rs

```diff
@@ -0,0 +1,312 @@
+            if ty.is_integral() || ty.is_bool() {
+                Ok(token)
+            } else {
+                Err((span, "only int and bool operations are stable in const fn".into()))
+            }
+        }
+        // checked by regular const fn checks
+        Rvalue::NullaryOp(..) => Ok(token),
+        Rvalue::UnaryOp(_, operand) => {
+            let ty = operand.ty(mir, tcx);
+            if ty.is_integral() || ty.is_bool() {
+                check_operand(tcx, mir, operand, span, token)
+            } else {
+                Err((span, "only int and bool operations are stable in const fn".into()))
+            }
+        }
+        Rvalue::Aggregate(_, operands) => {
+            let mut token = token;
+            for operand in operands {
+                token = check_operand(tcx, mir, operand, span, token)?;
+            }
+            Ok(token)
+        }
+    }
+}
+
+fn check_statement(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    mir: &'a Mir<'tcx>,
+    statement: &Statement<'tcx>,
+    token: IsMinConstFn,
+) -> McfResult {
+    let token = check_unsafety(mir, statement.source_info, token)?;
+    let span = statement.source_info.span;
+    match &statement.kind {
+        StatementKind::Assign(place, rval) => {
+            let token = check_place(tcx, mir, place, span, token)?;
+            check_rvalue(tcx, mir, rval, span, token)
+        }
+
+        StatementKind::ReadForMatch(_) => Err((span, "match in const fn is unstable".into())),
+
+        // just an assignment
+        StatementKind::SetDiscriminant { .. } => Ok(token),
+
+        | StatementKind::InlineAsm { .. }
+        => Err((span, "cannot use inline assembly in const fn".into())),
+
+        // These are all NOPs
+
+        | StatementKind::StorageLive(_)
+        | StatementKind::StorageDead(_)
+        | StatementKind::Validate(..)
+        | StatementKind::EndRegion(_)
+        | StatementKind::UserAssertTy(..)
+        | StatementKind::Nop
+        => Ok(token)
+    }
+}
+
+fn check_operand(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    mir: &'a Mir<'tcx>,
+    operand: &Operand<'tcx>,
+    span: Span,
+    token: IsMinConstFn,
+) -> McfResult {
+    match operand {
+        | Operand::Move(place)
+        | Operand::Copy(place)
+        => check_place(tcx, mir, place, span, token),
+        Operand::Constant(_) => Ok(token),
+    }
+}
+
+fn check_place(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    mir: &'a Mir<'tcx>,
+    place: &Place<'tcx>,
+    span: Span,
+    token: IsMinConstFn,
+) -> McfResult {
+    match place {
+        // locals are fine, because the `const_let` checks prevent them
+        | Place::Local(_)
+        // promoteds are always fine, they are essentially constants
+        | Place::Promoted(_)
+        => Ok(token),
+        Place::Static(_) => Err((span, "cannot access `static` items in const fn".into())),
+        Place::Projection(proj) => {
+            match proj.elem {
+                // raw ptr derefs prevented by `const_raw_ptr_deref`
```

Comment:
What is `const_raw_ptr_deref`?

Reply:
If it aint possible to detect this in MIR perhaps do it in HIR instead?

Reply:
AFAIK MIR has some unsafety information; we do some unsafety checks in MIR.

Reply:
done

Reply:
Done as in, the reference to `const_raw_ptr_deref` is gone? If not, please mention the filename where that function is defined...^^

Reply:
As in there's a test. And `const_raw_ptr_deref` is a feature gate.

Reply:
Oh, I see. That's not clear.^^

But the feature is ruled out independent of whether that gate is set...?

Reply:
well... it's additionally prevented in `min_const_fn` due to its unsafety. The same doesn't hold for casting `*const T` to `usize`. I'll add some more sanity checks. This will annoy the heck out of nightly users since they are now getting a lot of messages twice in different wording (already happening with a bunch of things in this PR)

Reply:
That's a consequence of doing multiple analyses, right?

I do not think that is the right approach... even if we have separate analyses, we should only call one.